### PR TITLE
Introduce apigen config/templates for generating documentation

### PR DESCRIPTION
**Pull request by EvanHerman**
This PR introduces a new config file for <a href="http://www.apigen.org">apigen</a> and template files for generating documentation for Primer.

Additionally, a new grunt task was registered `docs`, which clears the documentation directory and prepares it for new documentation files, and then generates the documentation by running `apigen` through shell.

Additional files may need to be added to the "Hook Docs" section to catch more actions/filters.

#### Package Overview
Package Docs

#### Class Overview
Class Docs

#### Action & Filter Docs
Actions & Filter Docs

### CODE REVIEW

File: apigen/hook-docs.php [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+	/**
+	 * Process files, pull hooks & render table
+	 *
+	 * @return mixed
+	 */
+	public static function process_hooks() {
+
+		$template_files   = self::get_files( '*.php', GLOB_MARK, '../templates/parts/' );
+
+		$other_files     = array(
```
Comment (jonathanbardo) [inline apigen/hook-docs.php]: @EvanHerman I think there is a spacing issue here.

File: apigen/hook-docs.php [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+
+				if ( in_array( self::$current_file, $scanned ) ) {
+
+					continue;
+
+				}
+
+				$scanned[] = self::$current_file;
+
+				foreach ( $tokens as $index => $token ) {
```
Comment (jonathanbardo) [inline apigen/hook-docs.php]: @EvanHerman Where was this taken from? The amount of indentation is pretty high here.

Comment (EvanHerman) [reply apigen/hook-docs.php]: @jonathanbardo I agree. it was originally a fork of https://github.com/woocommerce/woocommerce/blob/master/apigen/hook-docs.php which was extremely messy to begin with. I tried to clean it up as best I could without affecting the underlying functionality.

**Comment by frankiejarrett** [comment]
@EvanHerman I'd love to have a tool that could convert PHP docs into markdown and push it to GitHub pages automatically. Kind of similar to how the [WP-CLI site](https://github.com/wp-cli/wp-cli.github.com) works, except we'd need the ability to configure the structure on a per-project basis. Like with a `.docs.yml` or something. I'm not sure anything like this exists.

**Comment by EvanHerman** [comment]
@fjarrett I'm not sure one does either. These documentation files are standard HTML markup, so maybe we can convert from HTML -> Markdown and deploy to GitHub that way? It looks like the WP-CLI documentation is written in markdown and then converted to .html when the the docs are generated, but I haven't spent too much time digging in there.

**Comment by peterh44** [comment]
Awesome! Joined in the OnePercentBetter.

**Comment by frankiejarrett** [comment]
@EvanHerman Yeah at this point generating docs into Jekyll is probably a pipedream. I searched around the CakePHP docs (which uses ApiGen and looks amazing) to find a good example of how they're using markdown in the comments, and found this one:

**Docs**
https://api.cakephp.org/3.0/class-Cake.Cache.Cache.html

**Source**
https://github.com/cakephp/cakephp/blob/df1b9b8aa75ece32dbe5e5343a249f3058ba1f3f/src/Cache/Cache.php#L23-L67

**Comment by EvanHerman** [comment]
I went ahead and updated the repo with a new method of generating docs.

APIGen is still used to generate the code documentation, but additional pages are now written in either .md or .rst files and compiled using sphinx.

The documentation is now hosted on Github Pages for this repo:
https://godaddy.github.io/wp-primer-theme/index.html

**Comment by frankiejarrett** [comment]
😻

**Comment by frankiejarrett** [comment]
@EvanHerman Please resolve the conflict with `develop`

**Comment by EvanHerman** [comment]
@fjarrett Conflicts resolved

**Comment by frankiejarrett** [comment]
@EvanHerman @jonathanbardo Should all of this go inside `.dev/` since it's really not part of the theme package?

**Comment by EvanHerman** [comment]
@fjarrett Most likely. That sounds like a good spot for it. I wasn't quite sure where to place it since there were quite a lot of files.

**Comment by frankiejarrett** [comment]
@EvanHerman Cool, yeah I say let's move all this to `.dev/docs` and then I will merge.

File: Gruntfile.js [hunk 3]
```diff
@@ -232,10 +237,102 @@ module.exports = function( grunt ) {
...
+					func : 'writeln',
+					color: 'white',
+					styles: []
+				},
 			}
+		},
+
+		shell: {
+			sphinx: [
+				'cd .dev/docs',
```
Comment (frankiejarrett) [inline Gruntfile.js]: @EvanHerman what if we just add `easy_install pip` and `pip install -r requirements.txt` to the top of this array? macOS comes with Python 2.7, so checking for it really isn't necessary.

Comment (EvanHerman) [reply Gruntfile.js]: @fjarrett  I'll give it a test. I had to run sudo or else I was hitting permissions error. Running sudo in the shell task had a bug when it prompted for password, where typing just didn't work. 

I had assumed that we wouldn't be be the only ones making PRs on this, and Python isn't included on Windows (to my knowledge). To keep it succinct we can certainly include it in the grunt task.

File: .dev/docs/apigen/godaddy/@layout.latte [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	<link n:if="$config->googleCseId" rel="search" type="application/opensearchdescription+xml" title="{$config->title}" href="{$config->baseUrl}/opensearch.xml">
+
+	<script n:if="$config->googleAnalytics" type="text/javascript">
+		var _gaq = _gaq || [];
+		_gaq.push(['_setAccount', {$config->googleAnalytics}]);
+		_gaq.push(['_trackPageview']);
+
+		(function() {
+			var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
+			ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
```
Comment (frankiejarrett) [inline .dev/docs/apigen/godaddy/@layout.latte:45]: @EvanHerman Is this ours?

Comment (frankiejarrett) [reply .dev/docs/apigen/godaddy/@layout.latte:45]: @EvanHerman Oh never mind, I see it's dependent on a config!

**Pull request merged**